api.c

```diff
@@ -0,0 +1,434 @@
+    }
+
+    if (curr_state == InArg) {
+        arr_push(&out_args, curr_arg, &out_args_size, &out_args_length);
+    } else if (curr_state == InArgQuote) {
+        return 1;
+    }
+
+    size_t argc = out_args_length;
+    char **argv = arr_copy(out_args, argc);
+
+    free(out_args);
+    free(curr_arg);
+
+    data->argc = argc;
+    data->argv = argv;
+
+    return 0;
+}
+
+static void init_api_options(struct option longopts[]) {
+    int enabled_options[] = {
+        'c', 'u', 'i', 998, 't', 'C', 'F', 'L', 'M', 'B',
+        // Indicators colors
+        300, 301, 302, 303, 304, 305, 306, 307, 308, 309,
+        310, 311, 322, 312, 313, 314, 315,
+        // Text outline colors
+        316, 317, 318, 319, 320, 321, 323, 'r', 's',
+        'k', 400, 401, 402, 403,
+        // Alignment
+        500, 501, 502, 503, 504, 505, 506,
+        // String stuff
+        510, 511, 512, 513, 514, 515, 516, 517, 518, 519,
+        // Fonts
+        520, 521, 522, 523, 524, 525,
+        // Text size
+        530, 531, 532, 533, 534, 535, 536,
+        // Text/indicator position
+        540, 541, 542, 543, 544, 545, 546, 547, 548,
+        // Text outline width
+        560, 561, 562, 563, 564, 565, 566,
+        // Bar indicator stuff
+        700, 701, 703, 704, 705, 706, 707, 708, 709,
+        710, 711,
+
+        0
+    };
+    size_t size = sizeof(enabled_options)/sizeof(*enabled_options);
+    api_options = malloc(size * sizeof(struct option));
+
+    int n = 0;
+    int i = 0;
+
+    while (n < size) {
+        if (longopts[i].val == enabled_options[n]) {
+            api_options[n] = longopts[i];
+            n++;
+        }
+        i++;
+    }
+}
+
+void *listen_api(void* lopts) {
+    struct option *longopts = lopts;
+
+    if (access(api_fifo_path, F_OK) == 0) {
+        if (remove(api_fifo_path) != 0) {
+            fprintf(stderr, "Could not remove file \"%s\": %s\n",
+                    api_fifo_path, strerror(errno));
+            return NULL;
+        }
+    }
+    if (mkfifo(api_fifo_path, S_IRWXU) != 0) {
+        fprintf(stderr, "Could not create FIFO file \"%s\": %s\n",
+                api_fifo_path, strerror(errno));
+        return NULL;
+    }
+
+    FILE *fifo;
+    char c;
+    int i;
+    int times = 0;
+    char *text = malloc(50);
+    arg_data data;
+
+    DEBUG("Created fifo at \"%s\"\n", api_fifo_path);
+
+    init_api_options(longopts);
+
+#define open_fifo() {\
+    fifo = fopen(api_fifo_path, "rb");\
+    if (fifo == NULL) {\
+        fprintf(stderr, "Could not open FIFO file \"%s\": %s\n", \
+                api_fifo_path, strerror(errno));\
+        return NULL;\
+    }\
+}
+#define close_fifo() {\
+    if (fifo != NULL) fclose(fifo);\
+}
+    open_fifo()
+    bool reading = true;
+
+    while (1) {
+        strcpy(text, "command ");
+        i = strlen(text);
+        int n = i;
+
+        reading = true;
+        while (reading && (c = fgetc(fifo)) != EOF) {
+            if (sizeof(text) <= i) {
```

Comment:
sizeof doesn't work like this on malloc'd `text`. Might want to use a variable to record the size.